### PR TITLE
processResponseDone should receive a response

### PR DESCRIPTION
**Pull request by noamr**
Only call processResponseDone for responses that have their
timing info set, which are responses that have passed CORS
and redirects.

Closes https://github.com/whatwg/fetch/issues/1201

***
<a href="https://whatpr.org/fetch/1202.html" title="Last updated on Apr 21, 2021, 10:57 AM UTC (b455c91)">Preview</a> | <a href="https://whatpr.org/fetch/1202/6453947...b455c91.html" title="Last updated on Apr 21, 2021, 10:57 AM UTC (b455c91)">Diff</a>

### CODE REVIEW

**Comment by noamr** [comment]
@annevk

File: fetch.bs [hunk 1]
```diff
@@ -4019,7 +4027,7 @@ steps:
 
  <li><p>Set <var>response</var>'s <a for="response">timing info</a> to <var>timingInfo</var>.
 
- <li><p><a href="https://github.com/w3c/resource-timing/pull/261">Mark resource timing</a> for
+ <li><p><span>Mark resource timing</span> for
```
Comment (annevk) [inline fetch.bs:4030]: You want to use `<a>` here.

File: fetch.bs [hunk 2]
```diff
@@ -4251,9 +4259,20 @@ these steps:
...
-    <a>network error</a>.
+    <a>CORS check</a> for <var>request</var> and <var>response</var> returns failure, then:
+    <ul>
+     <li><p>Let <var>response</var> be a <a>network error</a>.
 
+     <li><p>Set <var>response</var>'s <a for=response>timing info</a> to a new
+     <a for=/>fetch timing info</a> with its <a for="fetch timing info">start time</a> and its
+     <a for="fetch timing info">post-redirect start time</a> set to <var>fetchParams</var>'s
+     <a for="fetch params">timing info</a>'s <a for="fetch timing info">start time</a>, and
+     its <a for="fetch timing info">end time</a> set to the <a for=/>coarsened shared current
```
Comment (annevk) [inline fetch.bs:4270]: No newlines inside `<a>` et al.

File: fetch.bs [hunk 3]
```diff
@@ -4251,9 +4259,20 @@ these steps:
 
    <li>
     <p>If <var>request</var>'s <a for=request>response tainting</a> is "<code>cors</code>" and a
-    <a>CORS check</a> for <var>request</var> and <var>response</var> returns failure, then return a
-    <a>network error</a>.
+    <a>CORS check</a> for <var>request</var> and <var>response</var> returns failure, then:
+    <ul>
```
Comment (annevk) [inline fetch.bs:4263]: This should be an ordered list.

**Review by annevk — COMMENTED** [review]
I think this is problematic as this would distinguish a CORS network error from other network errors. That they are indistinguishable is a desirable security property.

Can we still remove this from implementations?

I checked other network errors, such as a bad port, and we don't seem to be creating RT entries there, but I haven't done exhaustive testing for all possible network errors.

I'm also a bit concerned that somewhere we might do if "response is a network error, return a network error" which would drop this kind of information. I don't think we've audited that as there hasn't been a reason for it thus far.

cc @yoavweiss @npm1 @rniwa @bdekoz @sefeng211

**Comment by noamr** [comment]
> I think this is problematic as this would distinguish a CORS network error from other network errors. That they are indistinguishable is a desirable security property.
> 

All network errors that come after the CORS check should go through `fetch finale` and receive the RT entry, which only has a start/end time.
It's true that only CORS-or-after errors would receive an RT entry. Is that still too exposing?
I can remove this bit for now so that we can proceed with the other issues.

**Comment by noamr** [comment]
Added https://github.com/whatwg/fetch/issues/1215 to address the CORS/network error issue
This PR handles other issues.

**Comment by npm1** [comment]
> I think this is problematic as this would distinguish a CORS network error from other network errors. That they are indistinguishable is a desirable security property.

We expose errors in RT to prevent an attacker from distinguishing HTTP status codes (success vs errors). Would something similar apply here for CORS or not really (distinguishing success vs failed)?

**Comment by annevk** [comment]
I see where the confusion lies. A HTTP status code of 4xx, 5xx, or even 9xx is not a network error. That's a normal response as far as fetch is concerned. A network error is something more fundamental or the result of a security policy, e.g.:

* DNS error
* TLS error
* Broken H/2 framing
* CORS error
* CSP block
* Mixed Content block

Whether a response with a non-2xx HTTP status code results in an "error" depends on the endpoint. For `<object>` and `<script>` it does, for `` it does not. (I.e., unfortunately 2xx vs non-2xx is exposed to some extent.)

I think in principle we could do one of these:

* expose timing information for all of these, including network errors (using TAO as a policy and treating network errors as not having TAO).
* expose timing information for responses, but not network errors
* expose timing information for responses that result in a "load" event on their endpoint

But what we cannot do (in my opinion) is distinguish between network errors.

File: fetch.bs [hunk 4]
```diff
@@ -4017,9 +4032,10 @@ steps:
 
  <li><p>Set <var>response</var>'s <a for="response">timing info</a> to <var>timingInfo</var>.
 
- <li><p><a href="https://github.com/w3c/resource-timing/pull/261">Mark resource timing</a> for
- <var>timingInfo</var>, <var>originalURL</var>, <var>initiatorType</var>, and <var>global</var>.
- <!-- TODO -->
+ <li><p>
+ <a href="https://w3c.github.io/resource-timing/#dfn-mark-resource-timing">Mark resource timing</a>
```
Comment (annevk) [inline fetch.bs:4036]: Why is this needed, is it not exported? Let's export it first.

File: fetch.bs [hunk 5]
```diff
@@ -3981,13 +3981,21 @@ steps:
...
+ <li><p>Let <var>timingInfo</var> be <var>response</var>'s <a for=response>timing info</a>.
+
  <li><p>Set <var>fetchParams</var>'s <a for="fetch params">request</a>'s
  <a for=request>done flag</a>.
 
+ <li><p>If <var>timingInfo</var> is null, then return.
+
+ <li><p>Set <var>timingInfo</var>'s <a for="fetch timing info">end time</a> to
+ the <a for=/>coarsened shared current time</a> given <var>fetchParams</var>'s
+ <a for="fetch params">cross-origin isolated capability</a>.
```
Comment (annevk) [inline fetch.bs:3993]: Why do we need to set end time twice?

File: fetch.bs [hunk 6]
```diff
@@ -7217,8 +7233,9 @@ method steps are:
    <li><p><a lt=terminated for=fetch>Terminate</a> the ongoing fetch with the aborted flag set.
   </ol>
 
- <li><p>Let <var>handleFetchDone</var> be to <a>finalize and report timing</a> with
- <var>response</var>, <var>globalObject</var>, and "<code>fetch</code>".
+ <li><p>Let <var>handleFetchDone</var> given <a for=/>response</a> <var>response</var> be to
+ <a>finalize and report timing</a> with <var>response</var>, <var>globalObject</var>, and
+ "<code>fetch</code>".
```
Comment (annevk) [inline fetch.bs:7238]: I think you should rename the _response_ variable here as otherwise it's confusing with the other one, no?

File: fetch.bs [hunk 7]
```diff
@@ -4017,9 +4032,10 @@ steps:
 
  <li><p>Set <var>response</var>'s <a for="response">timing info</a> to <var>timingInfo</var>.
 
- <li><p><a href="https://github.com/w3c/resource-timing/pull/261">Mark resource timing</a> for
- <var>timingInfo</var>, <var>originalURL</var>, <var>initiatorType</var>, and <var>global</var>.
- <!-- TODO -->
+ <li><p>
+ <a href="https://w3c.github.io/resource-timing/#dfn-mark-resource-timing">Mark resource timing</a>
+ for <var>timingInfo</var>, <var>originalURL</var>, <var>initiatorType</var>, <var>global</var> and
```
Comment (annevk) [inline fetch.bs:4037]: Oxford comma.

File: fetch.bs [hunk 8]
```diff
@@ -4002,13 +4010,20 @@ steps:
...
+ <li><p>Let <var>cacheState</var> be <var>response</var>'s <a for=response>cache state</a>.
+
  <li><p>If <var>timingInfo</var> is null, then return.
 
- <li><p>If <var>response</var>'s <a for=response>timing allow passed flag</a> is not set, then set
- <var>timingInfo</var> to a new <a for=/>fetch timing info</a> whose
- <a for="fetch timing info">start time</a> and
- <a for="fetch timing info">post-redirect start time</a> are <var>timingInfo</var>'s
- <a for="fetch timing info">start time</a>.
+ <li><p>If <var>response</var>'s <a for=response>timing allow passed flag</a> is not set, then
```
Comment (annevk) [inline fetch.bs:4017]: As this `<li>` contains multiple flow elements the `<p>` needs to indented on a newline.

**Comment by annevk** [comment]
(I'm still okay with landing this modulo the above comments and leaving the error question to #1215.)

File: fetch.bs (hunk 4 above)
Comment (noamr) [reply fetch.bs:4036]: Sure, waiting for Resource-Timing to be published to TR (bikeshed vs. Respec stuff).

Comment (noamr) [reply fetch.bs:4036]: Removing this bit for now, will update once the refs are resolvable.

File: fetch.bs (hunk 5 above)
Comment (noamr) [reply fetch.bs:3993]: We don't actually. Removing.

File: fetch.bs (hunk 6 above)
Comment (noamr) [reply fetch.bs:7238]: ok, renaming to `responseObject`

File: fetch.bs (hunk 7 above)
Comment (noamr) [reply fetch.bs:4037]: ack

File: fetch.bs (hunk 8 above)
Comment (noamr) [reply fetch.bs:4017]: ack

**Comment by noamr** [comment]
> I see where the confusion lies. A HTTP status code of 4xx, 5xx, or even 9xx is not a network error. That's a normal response as far as fetch is concerned. A network error is something more fundamental or the result of a security policy, e.g.:
> 
> * DNS error
> * TLS error
> * Broken H/2 framing
> * CORS error
> * CSP block
> * Mixed Content block
> 
> Whether a response with a non-2xx HTTP status code results in an "error" depends on the endpoint. For `<object>` and `<script>` it does, for `` it does not. (I.e., unfortunately 2xx vs non-2xx is exposed to some extent.)
> 
> I think in principle we could do one of these:
> 
> * expose timing information for all of these, including network errors (using TAO as a policy and treating network errors as not having TAO).
> * expose timing information for responses, but not network errors
> * expose timing information for responses that result in a "load" event on their endpoint
> 
> But what we cannot do (in my opinion) is distinguish between network errors.

Let's continue the discussion in https://github.com/whatwg/fetch/issues/1215.

File: fetch.bs [hunk 9]
```diff
@@ -3981,13 +3981,17 @@ steps:
 <a for=/>response</a> <var>response</var>, run these steps:
 
 <ol>
+ <li><p>Let <var>timingInfo</var> be <var>response</var>'s <a for=response>timing info</a>.
+
  <li><p>Set <var>fetchParams</var>'s <a for="fetch params">request</a>'s
  <a for=request>done flag</a>.
 
+ <li><p>If <var>timingInfo</var> is null, then return.
```
Comment (annevk) [inline fetch.bs:3989]: Why do we have this here and also below in finalize and report timing? It seems this doesn't need to be here and you should get process response done either way?

File: fetch.bs [hunk 10]
```diff
@@ -4002,13 +4006,21 @@ steps:
...
 
- <li><p>If <var>response</var>'s <a for=response>timing allow passed flag</a> is not set, then set
- <var>timingInfo</var> to a new <a for=/>fetch timing info</a> whose
- <a for="fetch timing info">start time</a> and
- <a for="fetch timing info">post-redirect start time</a> are <var>timingInfo</var>'s
- <a for="fetch timing info">start time</a>.
+ <li>
+  <p>If <var>response</var>'s <a for=response>timing allow passed flag</a> is not set, then
+  perform the following steps:
+  <ol>
```
Comment (annevk) [inline fetch.bs:4016]: Newline before this.

File: fetch.bs [hunk 11]
```diff
@@ -4002,13 +4006,21 @@ steps:
...
+    <li><p>Set <var>timingInfo</var> to a new <a for=/>fetch timing info</a> whose
```
Comment (annevk) [inline fetch.bs:4017]: Single-space indentation.

File: fetch.bs [hunk 12]
```diff
@@ -4017,9 +4029,10 @@ steps:
 
  <li><p>Set <var>response</var>'s <a for="response">timing info</a> to <var>timingInfo</var>.
 
- <li><p><a href="https://github.com/w3c/resource-timing/pull/261">Mark resource timing</a> for
- <var>timingInfo</var>, <var>originalURL</var>, <var>initiatorType</var>, and <var>global</var>.
- <!-- TODO -->
```
Comment (annevk) [inline fetch.bs:4022]: Let's restore the TODO.

File: fetch.bs (hunk 6 above)
Comment (annevk) [reply fetch.bs:7235]: Oh sorry, this can be _response_ because _responseObject_ is used in the outer scope. And _response_ is used below, but only within the scope of those substeps.

File: fetch.bs [hunk 13]
```diff
@@ -1932,10 +1932,6 @@ message as HTTP/2 does not support them.
 <dfn export for=response id=concept-response-cache-state>cache state</dfn> (the empty string or
 "<code>local</code>"). Unlesss stated otherwise, it is the empty string.
 
-<p class=note>This is intended solely for usage by service workers. [[SW]]
-<!-- If we ever expand the utility of this we need to carefully consider whether filtered responses
-     need to mask it, whether the cache API needs to store it, etc. -->
```
Comment (annevk) [inline fetch.bs:1937]: I think we should keep this including the comment and instead say something like

> `This is intended for usage by <cite>Service Workers</cite> and <cite>Resource Timing</cite>.`

I guess for Resource Timing we also need opaque responses (that have TAO) to leak this information. I'll have a look if that needs more work.

File: fetch.bs (hunk 12 above)
Comment (noamr) [reply fetch.bs:4023]: Done

File: fetch.bs (hunk 13 above)
Comment (noamr) [reply fetch.bs:1937]: Done

**Comment by annevk** [comment]
@noamr I think it would be good if there was a corresponding PR to Resource Timing the commit could link to for it acknowledging the new _cacheState_ argument.

**Comment by noamr** [comment]
> @noamr I think it would be good if there was a corresponding PR to Resource Timing the commit could link to for it acknowledging the new _cacheState_ argument.

It’s already there in the editors draft.

**Comment by annevk** [comment]
I found the link above, it's https://github.com/w3c/resource-timing/pull/266.
